### DISCUSSION TITLE
Self-hosting

### DISCUSSION THREAD

**Discussion by mterenzio** [ampproject/bento#3] [General]
Is it possible to self-host the component libraries? If so, any advice or considerations?

**Comment by kristoferbaxter** [comment]
It should be possible to self host the libraries, but I'm going to tag in @danielrozenberg who has been working on the generic architecture for self-hosting overall.

He's (IIRC) working on a Cloudflare Worker hosting solution that adheres to all the auto-upgrade philosophies.

**Comment by samouri** [comment]
@mterenzio: can you explain more about what self-hosting means? It was a relevant concept for the AMP Runtime / AMP Cache, but I don't believe it applies to Bento. You should be able to grab the JS for Bento Web Components and stick them on any server you like.

**Comment by mterenzio** [comment]
Well, say for example, I'd like to use Bento components in a DAPP. I'd
rather not have a dependency on a domain bound CDN. I'd like the associated
libs to travel along with the web page into the environment

**Reply by mterenzio** [reply]
@samouri ^^

**Reply by danielrozenberg** [reply]
afaik (@samouri, @kristoferbaxter, correct me if I'm wrong) there's really nothing that prevents you from simply downloading the .js/.mjs files from cdn.ampproject.org and adding them to your own domain, though by doing so you'll lose the auto-update capabilities

If you have a build process that uses a `package.json` file you can add `@bentoproject/{component-name}` as a dependency and import it in your pre-compiled code: https://bentojs.dev/get-started/

This way you can use a dependency versioning manager (such as Renovate) to ensure you get code updates

We should clarify our documentation for the pros and cons of every approach but this is the "TL;DR" version of it

**Reply by samouri** [reply]
One correction, if you are trying to get _bento_ components you should not retrieve them from `cdn.ampproject.org`. The AMP Cache is for AMP Components. We publish Bento components to NPM, so you can retrieve them there or from any of its mirrors, i.e. [unpkg]( https://unpkg.com/browse/@bentoproject/fit-text@1.2111242025.1/dist/).

**Comment by mterenzio** [comment]
Thanks Daniel, makes sense
